pkg/arvo/app/chat-pull-hook.hoon

```diff
@@ -0,0 +1,260 @@
+    =/  tnk  p.sign
+    ?~  tnk  [~ this]
+    ?.  ?=([%backlog @ @ @ *] wire)  [~ this]
+    =/  chat=path  (oust [(dec (lent t.wire)) 1] `(list @ta)`t.wire)
+    :_  this
+    %.  :~  :*  %pass
+                /
+                %agent
+                [our.bol %chat-view]
+                %poke
+                %chat-view-action
+                !>([%delete chat])
+        ==  ==
+    %-  slog
+    :*  leaf+"chat-pull-hook failed subscribe on {(spud chat)}"
+        leaf+"stack trace:"
+        u.tnk
+    ==
+  ::
+      %fact
+    |^
+    ?+  p.cage.sign  (on-agent:def wire sign)
+        %chat-update    (fact-chat-update !<(update:store q.cage.sign))
+        %invite-update  (fact-invite-update !<(invite-update q.cage.sign))
+    ==
+    ::
+    ++  fact-chat-update
+      |=  =update:store
+      ^-  (quip card _this)
+      |^
+      ?+  -.update   [~ this]
+          %create
+        :_  this
+        ?>  ?=([* ^] path.update)
+        =/  shp  (~(get by tracking) path.update)
+        ?~  shp  ~
+        ?.  =(src.bol u.shp)  ~
+        [(chat-poke [%create path.update])]~
+      ::
+          %delete
+        ?>  ?=([* ^] path.update)
+        =/  shp  (~(get by tracking) path.update)
+        ?~  shp  [~ this]
+        ?.  =(u.shp src.bol)  [~ this]
+        =.  tracking  (~(del by tracking) path.update)
+        :_  this
+        :~  (chat-poke [%delete path.update])
+            :*  %pass
+                [%mailbox path.update]
+                %agent
+                [src.bol %chat-push-hook]
+                %leave
+                ~
+            ==
+            :*  %give
+                %fact
+                ~[/tracking]
+                %chat-pull-hook-update
+                !>([%tracking tracking])
+        ==  ==
+      ::
+          %message
+        :_  this
+        ?>  ?=([* ^] path.update)
+        =/  shp  (~(get by tracking) path.update)
+        ?~  shp  ~
+        ?.  =(src.bol u.shp)  ~
+        [(chat-poke [%message path.update envelope.update])]~
+      ::
+          %messages
+        :_  this
+        ?>  ?=([* ^] path.update)
+        =/  shp  (~(get by tracking) path.update)
+        ?~  shp  ~
+        ?.  =(src.bol u.shp)  ~
+        [(chat-poke [%messages path.update envelopes.update])]~
+      ==
+      ::
+      ++  chat-poke
+        |=  act=action:store
+        ^-  card
+        [%pass / %agent [our.bol %chat-store] %poke %chat-action !>(act)]
+      --
+    ++  fact-invite-update
+      |=  fact=invite-update
+      ^-  (quip card _this)
+      :_  this
+      ?+  -.fact  ~
+          %accepted
+        =/  ask-history  ?~((chat-scry:lib bol path.invite.fact) %.y %.n)
+        =*  shp       ship.invite.fact
+        =*  app-path  path.invite.fact
+        :~  :*  %pass
+                /
+                %agent
+                [our.bol %chat-view]
+                %poke
+                %chat-view-action
+                !>([%join shp app-path ask-history])
+        ==  ==
+      ==
+    --
+  ==
```

Comment:
This should be on the same column as the opening `|^`. Could be mobile formatting but it doesn’t look like it is.